components/eamxx/src/physics/mam/eamxx_mam_optics_process_interface.cpp

```diff
@@ -1,110 +1,552 @@
+#include <ekat/ekat_assert.hpp>
 #include <physics/mam/eamxx_mam_optics_process_interface.hpp>
 #include <share/property_checks/field_lower_bound_check.hpp>
 #include <share/property_checks/field_within_interval_check.hpp>
 
-#include "scream_config.h" // for SCREAM_CIME_BUILD
+#include "scream_config.h"  // for SCREAM_CIME_BUILD
+#include "share/grid/point_grid.hpp"
+#include "share/io/scorpio_input.hpp"
 
-#include <ekat/ekat_assert.hpp>
+namespace scream {
 
-namespace scream
-{
-
-MAMOptics::MAMOptics(
-    const ekat::Comm& comm,
-    const ekat::ParameterList& params)
-  : AtmosphereProcess(comm, params),
-    aero_config_() {
-}
+MAMOptics::MAMOptics(const ekat::Comm &comm, const ekat::ParameterList &params)
+    : AtmosphereProcess(comm, params), aero_config_() {
+ }
 
 AtmosphereProcessType MAMOptics::type() const {
   return AtmosphereProcessType::Physics;
 }
 
-std::string MAMOptics::name() const {
-  return "mam4_optics";
-}
+std::string MAMOptics::name() const { return "mam4_optics"; }
 
-void MAMOptics::set_grids(const std::shared_ptr<const GridsManager> grids_manager) {
+void MAMOptics::set_grids(
+    const std::shared_ptr<const GridsManager> grids_manager) {
   using namespace ekat::units;
 
-  grid_ = grids_manager->get_grid("Physics");
-  const auto& grid_name = grid_->name();
+  grid_                 = grids_manager->get_grid("Physics");
+  const auto &grid_name = grid_->name();
+  auto q_unit           = kg / kg;  // mass mixing ratios [kg stuff / kg air]
+  q_unit.set_string("kg/kg");
+  auto n_unit = 1 / kg;  // number mixing ratios [# / kg air]
+  n_unit.set_string("#/kg");
+  const auto m2 = m * m;
+  const auto s2 = s * s;
 
-  ncol_ = grid_->get_num_local_dofs();      // number of columns on this rank
-  nlev_ = grid_->get_num_vertical_levels(); // number of levels per column
-  nswbands_ = 14;                           // number of shortwave bands
-  nlwbands_ = 16;                           // number of longwave bands
+  ncol_     = grid_->get_num_local_dofs();  // number of columns on this rank
+  nlev_     = grid_->get_num_vertical_levels();  // number of levels per column
+  nswbands_ = mam4::modal_aer_opt::nswbands;     // number of shortwave bands
+  nlwbands_ = mam4::modal_aer_opt::nlwbands;     // number of longwave bands
 
   // Define the different field layouts that will be used for this process
   using namespace ShortFieldTagsNames;
 
   // Define aerosol optics fields computed by this process.
   auto nondim = Units::nondimensional();
-  FieldLayout scalar3d_swband_layout { {COL, SWBND, LEV}, {ncol_, nswbands_, nlev_} };
-  FieldLayout scalar3d_lwband_layout { {COL, LWBND, LEV}, {ncol_, nlwbands_, nlev_} };
+  // 3D layout for shortwave aerosol fields: columns, number of shortwave, and nlev +1 s
+  // shortwave aerosol fields in mam has one extra level (nlev+1)
+  FieldLayout scalar3d_swband_layout_ilev{{COL, SWBND, ILEV},
+                                      {ncol_, nswbands_, nlev_ + 1}};
```

Comment:
That's a good point. I replaced `_int` with `_ilev` to avoid confusion but you are right, we should perhaps name it differently. `ilevp1` is a good suggestion. Thanks!